ipa-core/src/protocol/ipa_prf/validation_protocol/validation.rs

```diff
@@ -213,6 +219,55 @@ impl BatchToVerify {
             challenges_for_prover_right.collect(),
         )
     }
+
+    /// This function computes and outputs the final `p_r_from_left_prover * q_r_from_left_prover` value.
+    async fn p_and_q_r_check<C, U, V>(
+        &self,
+        ctx: &C,
+        challenges_for_left_prover: &[Fp61BitPrime],
+        challenges_for_right_prover: &[Fp61BitPrime],
+        u_from_right_prover: U,
+        v_from_left_prover: V,
+    ) -> Result<Fp61BitPrime, Error>
+    where
+        C: Context,
+        U: Iterator<Item = Fp61BitPrime> + Send,
+        V: Iterator<Item = Fp61BitPrime> + Send,
+    {
+        const LRF: usize = LargeProofGenerator::RECURSION_FACTOR;
+        const SRF: usize = SmallProofGenerator::RECURSION_FACTOR;
+
+        // compute p_r
+        let p_r_right_prover = recursively_compute_final_check::<_, _, LRF, SRF>(
```

Comment:
I really think this is backwards... if the prover is to the "right" of this party, to me that means that this is party P_{i-1} for the prover P_i. And according to the paper, this verifier should be computing q(r).